File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/voicenote/VoiceNoteEntity.kt
/*
 * VoiceNoteEntity.kt
 *
 * Room @Entity for the `voice_notes` table. One row per voice note
 * recorded on a document's detail screen — the clip itself lives in
 * AttachmentStorage as an .m4a file and the row points at it through
 * `audio_uri`. A capture can have any number of voice notes;
 * deleting the capture cascades to its notes via the foreign key.
 *
 * Schema mirrors iOS GRDB `v12_voice_notes` migration byte-for-byte
 * so the Drive sync payload round-trips between platforms.
 *
 * Two drive-id columns:
 *   - `drive_file_id`        — JSON metadata on Drive.
 *   - `audio_drive_file_id`  — uploaded .m4a binary.
 * Split so the sync worker can update the transcript without re-
 * uploading the audio, the common case after first sync.
 */

package app.quickink.mobile.data.voicenote

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import app.quickink.mobile.data.capture.CaptureEntity

@Entity(
    tableName = "voice_notes",
    foreignKeys = [
        ForeignKey(
            entity        = CaptureEntity::class,
            parentColumns = ["id"],
            childColumns  = ["capture_id"],
            onDelete      = ForeignKey.CASCADE,
        ),
    ],
    indices = [
        Index(value = ["capture_id", "created_at"], name = "idx_voice_notes_capture"),
        Index(value = ["user_id", "created_at"],    name = "idx_voice_notes_user"),
        Index(value = ["dirty"],      name = "idx_voice_notes_dirty"),
        Index(value = ["deleted_at"], name = "idx_voice_notes_tombstone"),
    ],
)
data class VoiceNoteEntity(

    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "capture_id")
    val captureId: String,

    @ColumnInfo(name = "user_id")
    val userId: String,

    /** `file://` URL of the on-disk .m4a (AAC, 16kHz mono, 96kbps). */
    @ColumnInfo(name = "audio_uri")
    val audioUri: String,

    /** Committed clip duration in milliseconds. */
    @ColumnInfo(name = "duration_ms")
    val durationMs: Long,

    /** Speech-to-text result. Null until the user taps Transcribe. */
    @ColumnInfo(name = "transcription")
    val transcription: String? = null,

    /**
     * Which backend produced the transcript — one of
     * `SpeechTranscriber.BACKEND_MLKIT` / `BACKEND_SHERPA`, or
     * `"sfspeech"` for transcripts that came in from iOS via Drive.
     */
    @ColumnInfo(name = "transcription_source")
    val transcriptionSource: String? = null,

    /** Drive id for the JSON metadata row. NULL until first push. */
    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String? = null,

    /** Drive id for the uploaded .m4a. NULL until first push. */
    @ColumnInfo(name = "audio_drive_file_id")
    val audioDriveFileId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,
)

File: apps/quickink/ios/QuickInk/Data/VoiceNoteEntity.swift
/*
 * VoiceNoteEntity.swift
 *
 * One row per voice note attached to a capture. Voice notes are
 * captured on the Document detail screen — the user taps "Voice
 * notes" → "Record", talks for up to two minutes, and the resulting
 * .m4a lands in AttachmentStorage with a row pointing at it. A
 * capture can have any number of voice notes; deleting the capture
 * cascades to its voice notes via the `ON DELETE CASCADE` foreign
 * key (mirror of how `ocr_results` rows attach to captures).
 *
 * Storage layout:
 *   - `audio_uri`     — file:// URL of the on-disk .m4a. AAC, 44.1kHz
 *                       mono, 96kbps. Cleaned up when the row is
 *                       hard-deleted.
 *   - `duration_ms`   — committed clip duration in milliseconds.
 *   - `transcription` — speech-to-text result. Nullable while we're
 *                       still recording or transcribing fails.
 *   - `drive_file_id` — Drive id of the uploaded .m4a after sync.
 *                       NULL until the next sync push completes.
 *
 * Mirror of Android's `VoiceNoteEntity` (Room) — column names match
 * byte-for-byte so the Drive sync payload round-trips identically
 * between platforms.
 */

import Foundation
import GRDB

public struct VoiceNoteEntity: Codable, FetchableRecord, MutablePersistableRecord, Equatable {

    public var id: String
    public var captureId: String
    public var userId: String
    public var audioUri: String
    public var durationMs: Int
    public var transcription: String?
    public var transcriptionSource: String?
    public var driveFileId: String?
    public var audioDriveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public static var databaseTableName: String { "voice_notes" }

    public enum Columns {
        public static let id = Column(CodingKeys.id)
        public static let captureId = Column(CodingKeys.captureId)
        public static let userId = Column(CodingKeys.userId)
        public static let audioUri = Column(CodingKeys.audioUri)
        public static let durationMs = Column(CodingKeys.durationMs)
        public static let transcription = Column(CodingKeys.transcription)
        public static let transcriptionSource = Column(CodingKeys.transcriptionSource)
        public static let driveFileId = Column(CodingKeys.driveFileId)
        public static let audioDriveFileId = Column(CodingKeys.audioDriveFileId)
        public static let createdAt = Column(CodingKeys.createdAt)
        public static let updatedAt = Column(CodingKeys.updatedAt)
        public static let dirty = Column(CodingKeys.dirty)
        public static let deletedAt = Column(CodingKeys.deletedAt)
    }

    enum CodingKeys: String, CodingKey {
        case id
        case captureId           = "capture_id"
        case userId              = "user_id"
        case audioUri            = "audio_uri"
        case durationMs          = "duration_ms"
        case transcription
        case transcriptionSource = "transcription_source"
        case driveFileId         = "drive_file_id"
        case audioDriveFileId    = "audio_drive_file_id"
        case createdAt           = "created_at"
        case updatedAt           = "updated_at"
        case dirty
        case deletedAt           = "deleted_at"
    }
}
